pkg/controllers/provisioning/scheduling/topologynodefilter.go

```diff
@@ -28,43 +28,66 @@ import (
 // included for topology counting purposes. This is only used with topology spread constraints as affinities/anti-affinities
 // always count across all nodes. A nil or zero-value TopologyNodeFilter behaves well and the filter returns true for
 // all nodes.
-type TopologyNodeFilter []scheduling.Requirements
+type TopologyNodeFilter struct {
+	Requirements   []scheduling.Requirements
+	TaintPolicy    v1.NodeInclusionPolicy
+	AffinityPolicy v1.NodeInclusionPolicy
+	Tolerations    []v1.Toleration
+}
 
-func MakeTopologyNodeFilter(p *v1.Pod) TopologyNodeFilter {
+func MakeTopologyNodeFilter(p *v1.Pod, taintPolicy v1.NodeInclusionPolicy, affinityPolicy v1.NodeInclusionPolicy) TopologyNodeFilter {
 	nodeSelectorRequirements := scheduling.NewLabelRequirements(p.Spec.NodeSelector)
 	// if we only have a label selector, that's the only requirement that must match
 	if p.Spec.Affinity == nil || p.Spec.Affinity.NodeAffinity == nil || p.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
-		return TopologyNodeFilter{nodeSelectorRequirements}
+		return TopologyNodeFilter{
+			Requirements:   []scheduling.Requirements{nodeSelectorRequirements},
+			TaintPolicy:    taintPolicy,
+			AffinityPolicy: affinityPolicy,
+			Tolerations:    p.Spec.Tolerations,
+		}
 	}
 
 	// otherwise, we need to match the combination of label selector and any term of the required node affinities since
 	// those terms are OR'd together
-	var filter TopologyNodeFilter
+	filter := TopologyNodeFilter{
```

Comment:
We don't attach tolerations in this codepath? Seems wrong that we wouldn't attach tolerations here -- also, since there isn't a test failing for this scenario seems like we need to add one

Reply:
Good catch, looks like we need a test that involves pods with node selectors and an honor taint TSC.